Separate ns variable in eval_code into globals and locals

hoodmane opened the pull request:
This will allow `pyodide.js` to avoid pollution of global namespace by using `eval_code` with new empty dict.

casatir commented:
+1 for this. As a side effect, it replaces the `ns` parameter which I found not really explicit.

casatir commented on src/pyodide-py/pyodide/_base.py:
```diff
@@ -287,12 +290,19 @@ def eval_code(
     """
-    return CodeRunner(ns, mode, quiet_trailing_semicolon, filename).run(code)
+    return CodeRunner(
+        globals=globals,
+        locals=locals,
+        mode=mode,
+        quiet_trailing_semicolon=quiet_trailing_semicolon,
+        filename=filename,
+    ).run(code)
 
 
-async def _eval_code_async(
+async def eval_code_async(
```
One micro step towards official async support?

hoodmane replied:
Yeah.

casatir commented:
> avoid pollution of global namespace

Can you add some tests that demonstrate this please?

hoodmane commented:
> Can you add some tests that demonstrate this please?

Of course =)

hoodmane commented:
@casatir How's that look for a test?

rth commented on src/pyodide-py/pyodide/_base.py:
```diff
@@ -81,12 +81,14 @@ class CodeRunner:
         self,
-        ns: Optional[Dict[str, Any]] = None,
+        globals: Optional[Dict[str, Any]] = None,
+        locals: Optional[Dict[str, Any]] = None,
```
You are shadowing the `global` / `locals` objects here, which is not great. Maybe `ns_globals` and `ns_locals`?
You also need to update the docstring above and in `eval_*`

hoodmane replied:
Yeah, it's not great. I've seen this done in enough other peoples' code that I've gotten used to it. `ns_globals` seems fine to me. @casatir?

hoodmane replied:
@rth Actually, look at the built in `eval` and `exec` functions:
https://docs.python.org/3/library/functions.html#eval
https://docs.python.org/3/library/functions.html#exec
They also call these arguments `globals` and `locals`. I think it's a good idea to name them `globals` and `locals` here too.

rth reviewed (commented):
I agree it very useful. Please add a changelog entry.

rth commented on src/pyodide-py/pyodide/_base.py:
```diff
@@ -237,21 +242,22 @@ async def run_async(self, code: str) -> Any:
 def eval_code(
     code: str,
-    ns: Dict[str, Any],
+    globals: Dict[str, Any] = None,
+    locals: Dict[str, Any] = None,
```
```suggestion
    globals: Optional[Dict[str, Any]] = None,
    locals: Optional[Dict[str, Any]] = None,
```

rth commented on src/pyodide-py/pyodide/_base.py:
```diff
@@ -262,8 +268,12 @@ def eval_code(
+        See [the exec documentation](https://docs.python.org/3/library/functions.html#exec) for more info.
+    locals
+        The local scope in which to execute code. This is used as the `eval` `globals` parameter.
+        See [the exec documentation](https://docs.python.org/3/library/functions.html#exec) for more info.
```
Please wrap docstrings to 88 char as well.

rth commented on src/pyodide-py/pyodide/_base.py:
```diff
@@ -49,9 +49,12 @@ class CodeRunner:
+        See [the exec documentation](https://docs.python.org/3/library/functions.html#exec) for more info.
+    locals
+        The local scope in which to execute code. This is used as the `eval` `globals` parameter.
+        See [the exec documentation](https://docs.python.org/3/library/functions.html#exec) for more info.
```
Same

rth commented on src/pyodide-py/pyodide/_base.py:
```diff
@@ -49,9 +49,12 @@ class CodeRunner:
+        The global scope in which to execute code. This is used as the `exec` `globals` parameter.
+        See [the exec documentation](https://docs.python.org/3/library/functions.html#exec) for more info.
+    locals
+        The local scope in which to execute code. This is used as the `eval` `globals` parameter.
```
`eval` `locals` parameter. ?

Also indicate that when not provided `globals` parameter is used.

rth reviewed (commented):
A few remaining comments, otherwise LGTM. Thanks.

Also please add a changelog entry.

hoodmane commented:
I also changed `mode` ==> `return_mode` (more explicit) and changed it so that it will raise an error if passed an unexpected mode. Otherwise it's way too easy for someone to pass i.e., `"last-expr-or-assign"` or `"last_exp_or_assign"` and be confused when it does something different than expected. Could be friendly to offer:
```
class ReturnMode:
    LAST_EXPR = "last_expr"
    LAST_EXPR_OR_ASSIGN = "last_expr_or_assign"
    NONE = None
```
